README.md
# raven-nette

This is nette implementation of sentry raven-php as logger service for tracy.

Tested with nette 2.3>, please report any bugs into Issues

PR's are welcomed!

## Installation and usage

### Installation via composer:

```bash
composer require salamek/raven-nette
```

### Usage

Register extension to your config.neon:

```yaml
extensions:
  sentryLogger: Salamek\RavenNette\DI\SentryLoggerExtension
```

And configure by setting:

```yaml
sentryLogger:
  dsn: 'YOUR_SENTRY_DSN'

  # Optional configuration values
  inDebug: false # bool: Log in debug mode ? default is false
  directory: null # string|null: Where to store log files ? default is Debugger::$logDirectory, null to disable
  email: null # string|null :Where to send email notifications ? default is Debugger::$email, null to disable
  options: [release: YOUR_RELEASE] # array :All options supported by getsentry/sentry-php
  context:
    user: true # Send logged in user information
```
List of all confuration options for [getsentry/sentry-php](https://github.com/getsentry/sentry-php#configuration)

### Alternative Usage

If you dont want to use DI, and/or be able to log errors as soon as posible use this approach

Put this code into your `app/bootstrap.php` after RobotLoader is initiated and `$configurator->enableDebugger` is called:

```php
// Initiate sentryLogger
new \Salamek\RavenNette\SentryLogger(
  'YOUR_SENTRY_DSN', //Sentry DSN
  false, //Log in DEBUG mode ? //You dont want that...
  null, //Set where do you want to store file log (Tracy\Debugger::$logDirectory | null | string)
  null, //Send email as usual logger ?   (Tracy\Debugger::$email | null | string | array )
  true,
  ['release' => 'YOUR_RELEASE'] //All options supported by getsentry/sentry-php
);
```

### Usage only with tracy

If you dont want use nette at all but only raven-nette and tracy... well you can!

```php
include('vendor/autoload.php');
use Tracy\Debugger;

Debugger::enable(Debugger::PRODUCTION);

new \Salamek\RavenNette\SentryLogger(
  'YOUR_SENTRY_DSN', //Sentry DSN
  false, //Log in DEBUG mode ? //You dont want that...
  null, //Set where do you want to store file log (Tracy\Debugger::$logDirectory | null | string)
  null, //Send email as usual logger ?   (Tracy\Debugger::$email | null | string | array )
  true,
  ['release' => 'YOUR_RELEASE'] //All options supported by getsentry/sentry-php
);

Debugger::log('My error', 'error');
```

And that should be everything...


CHANGELOG.md
v1.3.1
------
* Switch from deprecated raven/raven to sentry/sentry

v1.3.0
------
* Option to send logged in user information to sentry when enabled in config [marten-cz](https://github.com/marten-cz)
* Added support for configuration options [xdobro4](https://github.com/xdobro4)

v1.2.0
------
* Added DI Extension
* Changed namespace from `Salamek` to `Salamek\RavenNette`

v1.1.1
------
Fixes issue when all Tracy events was send to Sentry, now only events of ERROR, EXCEPTION, CRITICAL and WARNING are send to Sentry (ignores DEBUG, INFO)

v1.1
----
First stable relase,
Removed need of manual connecting events from Tracy to raven-nette, so no need to use this code (this code is run in raven-nette __construct):

```
 // Add Fatal Error handler		
 \Tracy\Debugger::$onFatalError[] = function($e) use($sentryLogger)		
 {		
   $sentryLogger->onFatalError($e);		
 };		
 		
 // Add logger to tracy		
 Tracy\Debugger::setLogger($sentryLogger);
```

v1.0-alpha
----------
First testing release


src/SentryLogger.php
<?php
/*
 * Copyright (C) 2015 Adam Schubert <adam.schubert@sg1-game.net>.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 3.0+ of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301  USA
 *
 * OR
 *
 * Copyright (C) 2015, Adam Schubert
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * * Neither the name of test nor the names of its
 *  contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

namespace Salamek\RavenNette;

use Exception;
use Tracy\Debugger;
use Tracy\Logger;

/**
 * Description of sentryLogger
 *
 * @author Adam Schubert <adam.schubert@sg1-game.net>
 */
class SentryLogger extends Logger
{
    /** @var \Raven_Client */
    private $raven;

    /** @var bool */
    private $enabled = true;

    /**
     * SentryLogger constructor.
     * @param $dsn
     * @param bool $inDebug
     * @param null $directory
     * @param null $email
     * @param bool $autoWire
     * @param array $options
     */
    public function __construct($dsn, $inDebug = false, $directory = null, $email = null, $autoWire = true, $options = [])
    {
        parent::__construct($directory, $email, Debugger::getBlueScreen());

        //Check for production mode, you will want to fllod sentry only in production... right ?
        $this->enabled = Debugger::$productionMode || $inDebug;

        $this->raven = new \Raven_Client($dsn, $options);

        if ($autoWire) {
            //Add sentryNetteLogger to tracy
            $that = $this;
            Debugger::$onFatalError[] = function ($e) use ($that) {
                $that->onFatalError($e);
            };

            // Add logger to tracy
            Debugger::setLogger($this);
        }
    }

    /**
     * Set logged in user into raven context
     *
     * @param null $userId
     * @param null $email
     * @param array|NULL $data
     *
     * @return null
     */
    public function setUserContext($userId = null, $email = null, array $data = null)
    {
        $this->raven->set_user_data($userId, $email, $data);
    }

    /**
     * @param $message
     * @param string $priority
     * @return null|string
     */
    public function log($message, $priority = self::INFO)
    {
        if ($this->enabled) {
            $exceptionFile = '';
            if ($this->directory && is_dir($this->directory)) {
                $exceptionFile = $message instanceof Exception ? $this->getExceptionFile($message) : null;
                $line = $this->formatLogLine($message, $exceptionFile);
                $file = $this->directory . '/' . strtolower($priority ?: self::INFO) . '.log';

                if (!@file_put_contents($file, $line . PHP_EOL, FILE_APPEND | LOCK_EX)) {
                    throw new \RuntimeException("Unable to write to log file '$file'. Is directory writable?");
                }
            }

            if ($message instanceof Exception) {
                $this->raven->captureException($message);

                if ($this->directory && is_dir($this->directory)) {
                    $this->logException($message, $exceptionFile);
                }
            } else {
                if (in_array($priority, array(self::ERROR, self::EXCEPTION, self::CRITICAL, self::WARNING), true)) {
                    $this->raven->captureMessage($message, array(), $priority);
                }
            }

            if (in_array($priority, array(self::ERROR, self::EXCEPTION, self::CRITICAL), true)) {
                $this->sendEmail($message);
            }

            return $exceptionFile;
        } else {
            return parent::log($message, $priority);
        }
    }

    /**
     * @param $e
     */
    public function onFatalError($e)
    {
        if ($this->enabled) {
            $this->raven->captureException($e);
        }
    }
}


src/deprecated/sentryNetteLogger.php
<?php
/*
 * Copyright (C) 2015 Adam Schubert <adam.schubert@sg1-game.net>.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 3.0+ of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301  USA
 *
 * OR
 *
 * Copyright (C) 2015, Adam Schubert
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * * Neither the name of test nor the names of its
 *  contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


namespace Salamek;

use Salamek\RavenNette\SentryLogger;

/**
 * Description of sentryNetteLogger
 * @deprecated
 * @author Adam Schubert <adam.schubert@sg1-game.net>
 */
class sentryNetteLogger extends SentryLogger
{
    /**
     * sentryNetteLogger constructor.
     * @param $dsn
     * @param bool $inDebug
     * @param null $directory
     * @param null $email
     */
    public function __construct($dsn, $inDebug = false, $directory = null, $email = null)
    {
        trigger_error('Salamek\sentryNetteLogger is deprecated; use Salamek\RavenNette\SentryLogger instead.', E_USER_DEPRECATED);
        parent::__construct($dsn, $inDebug, $directory, $email, true);
    }
}


src/DI/SentryLoggerExtension.php
<?php

/*
 * Copyright (C) 2015 Adam Schubert <adam.schubert@sg1-game.net>.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 3.0+ of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301  USA
 *
 * OR
 *
 * Copyright (C) 2015, Adam Schubert
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * * Neither the name of test nor the names of its
 *  contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

namespace Salamek\RavenNette\DI;

use Salamek\RavenNette\SentryLogger;
use Nette\DI\CompilerExtension;
use Nette\PhpGenerator\ClassType;
use Nette\Utils\Validators;
use Tracy\Debugger;

/**
 * SentryLoggerExtension for Nette DI
 *
 * @author Adam Schubert <adam.schubert@sg1-game.net>
 */
class SentryLoggerExtension extends CompilerExtension
{
    /**
     * @param ClassType $class
     * @throws \Nette\Utils\AssertionException
     */
    public function afterCompile(ClassType $class)
    {
        $defaults = array();
        $defaults['inDebug'] = false;
        $defaults['directory'] = Debugger::$logDirectory;
        $defaults['email'] = Debugger::$email;
        $defaults['options'] = [];

        $config = $this->getConfig($defaults);

        Validators::assertField($config, 'dsn', 'string');

        $init = $class->getMethod('initialize');
        $init->addBody(
            '$sentryLogger = new ' . SentryLogger::class . '(?, ?, ?, ?, ?, ?);' .
            Debugger::class . '::$onFatalError[] = function($e) use($sentryLogger)' .
            '{' .
            '  $sentryLogger->onFatalError($e);' .
            '};' .
            Debugger::class . '::setLogger($sentryLogger);',
            array(
                $config['dsn'],
                $config['inDebug'],
                $config['directory'],
                $config['email'],
                false,
                $config['options']
            )
        );

        if (isset($config['context']['user']) && isset($config['context']['user'])) {
            $init->addBody(
                '$user = $this->getService(?);' .
                'if ($user->isLoggedIn()) { $sentryLogger->setUserContext($user->getId(), \'\', (array) $user->getIdentity()); }',
                ['user']
            );
        }
    }
}
